sycl/include/sycl/queue.hpp

```diff
@@ -117,10 +118,85 @@ class __SYCL_EXPORT queue {
   queue(const async_handler &AsyncHandler, const property_list &PropList = {})
       : queue(default_selector(), AsyncHandler, PropList) {}
 
+#if __cplusplus >= 201703L
+  /// Constructs a SYCL queue instance using the device identified by the
+  /// device selector provided.
+  /// \param DeviceSelector is SYCL 2020 Device Selector, a simple callable that
+  /// takes a device and returns an int
+  /// \param AsyncHandler is a SYCL asynchronous exception handler.
+  /// \param PropList is a list of properties for queue construction.
+  template <
+      typename DeviceSelector,
+      typename = std::enable_if_t<
+          std::is_invocable_r<int, DeviceSelector &, device &>::value &&
+          !std::is_base_of_v<ext::oneapi::filter_selector, DeviceSelector>>>
+  // see [filter_selectors not "Callable"] in device_selectors.cpp
+  explicit queue(const DeviceSelector &deviceSelector,
+                 const async_handler &AsyncHandler,
+                 const property_list &PropList = {})
+      : queue(detail::select_device(deviceSelector), AsyncHandler, PropList) {}
```

Comment:
```suggestion
      : queue(device { deviceSelector }, AsyncHandler, PropList) {}
```
?

Reply:
I like this suggestion, and if you prefer it I'll change it.

However, consider that of the four constructors we are adding here, only two are amenable to the change. The other two queue constructors use continue to use  `detail::select_device` because they have the additional context argument. Personally, I prefer all four to be consistent with each other, each calling detail::select_device directly, rather than just passing it to the device constructor in two cases.  This may simply be a matter of taste.  Let me know.

Reply:
Interesting you need this `detail::select_device` at the first place.
Reading again the specification on queue, `queue` creation throws if the device picked by the selector is not in the context. The device is not the best one picked from the devices in the context but it is picked independently. So it is unclear why this `detail::select_device` takes all these parametersn

Reply:
My reading was that the device selector would be used to select the best device _from_ the context (assuming an acceptable one could be found).  It makes little sense to even provide that constructor otherwise.

FWIW, that is how the analog SYCL 1.2.1 queue constructor is implemented right now, and I used its behavior to model this one ( https://github.com/intel/llvm/blob/sycl/sycl/source/queue.cpp#L23-L30 )

Reply:
Interesting feature, but this is not how the SYCL specification is defined:

> Constructs a SYCL queue instance that is associated with the syclContext provided, using the device returned by the [device selector](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#device-selector) provided. Must throw an exception with the errc::invalid error code if syclContext does not encapsulate the SYCL device returned by deviceSelector.

So either the specification has to be changed or DPC++ has to be updated. @bader any direction?
But anyway, this is outside of this PR.

Reply:
I agree with @keryell that device selection is supposed to be "context-free". I don't know if logic currently implemented by DPC++ is used, but it can be implemented outside of the library.
To honest, I don't have enough data to recommend any direction here. 
At first glace, DPC++ behavior sounds like a user friendly approach. The runtime emits less exceptions for the user to handle and tries to do it best to select the right device. On the other hand, it's just a syntax sugar, which might add an overhead and not sure if it's really used in the real word.
@intel/dpcpp-specification-reviewers, do you have any thoughts on that subject?

Reply:
I looked at all the places in the SYCL 2020 spec where a device selector is passed.  In all other cases, it's clear that the implementation should pass every root device to the device selector, and the selector ranks each one.

This queue constructor is the only place where the set of valid device is limited by a context that the user passes in.  It could make sense to redefine the spec to say that the implementation only passes devices from the context to the selector in this case.  However, that is not how the spec is worded now, and we have the same wording in SYCL 1.2.1.

I actually wonder if this queue constructor is even useful.  If the user is passing a custom context, maybe we should also force them to pass a specific device (and not a device selector).  This avoids any ambiguity about which devices the implementation passes to the selector.  We already have a form of the queue constructor like this (taking a context and a device). If we were doing this from scratch, I think I would not have added a form of the queue constructor that takes a context and a selector.

My feeling is that DPC++ should implement the existing spec, and maybe the committee should consider deprecating this form of the queue constructor.